src/main/java/org/kairosdb/core/http/rest/MetricsResource.java

```diff
@@ -439,65 +447,68 @@ public Response get(@Context HttpHeaders httpHeaders, String json) throws Except
 		checkNotNull(json);
 		logger.debug(json);
 
-		Span span = createSpan("datapoints_query", httpHeaders);
+		final Span span = createSpan("datapoints_query", httpHeaders);
+		ExecutorService executor = new ThreadPoolExecutor(0,4,0l,TimeUnit.MILLISECONDS, new SynchronousQueue<>(),
+				new ThreadPoolExecutor.AbortPolicy());
 
 		try (Scope scope = tracer.scopeManager().activate(span, false))
 		{
+			TimeLimiter limiter = new SimpleTimeLimiter(executor);
+
 			File respFile = File.createTempFile("kairos", ".json", new File(datastore.getCacheDir()));
 			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(respFile), "UTF-8"));
-
 			JsonResponse jsonResponse = new JsonResponse(writer);
-
 			jsonResponse.begin();
-
 			List<QueryMetric> queries = queryParser.parseQueryMetric(json);
 
-			for (QueryMetric query : queries)
-			{
-				Map<String, Collection<String>> tags = query.getTags().asMap();
-				Set<String> keys = tags.keySet();
-
-				span.setTag("metric_name", query.getName());
-
-				for (String key : keys) {
-					StringBuilder str = new StringBuilder();
-					Iterator<String> iter = tags.get(key).iterator();
-					while(iter.hasNext()) {
-						str.append(iter.next() + ",");
-					}
-					str.deleteCharAt(str.lastIndexOf(","));
-					span.setTag(key, str.toString());
-				}
-
-				queryMeasurementProvider.measureSpanForMetric(query);
-				queryMeasurementProvider.measureDistanceForMetric(query);
-
-				DatastoreQuery dq = datastore.createQuery(query);
-				try {
-					List<DataPointGroup> results = dq.execute();
-					jsonResponse.formatQuery(results, query.isExcludeTags(), dq.getSampleSize());
-				} catch (Throwable e) {
-					queryMeasurementProvider.measureSpanError(query);
-					queryMeasurementProvider.measureDistanceError(query);
-					throw e;
-				}
-				finally
-				{
-					queryMeasurementProvider.measureSpanSuccess(query);
-					queryMeasurementProvider.measureDistanceSuccess(query);
-					dq.close();
-				}
-			}
-
-			jsonResponse.end();
-			writer.flush();
-			writer.close();
-
-			ResponseBuilder responseBuilder = Response.status(Response.Status.OK).entity(
-					new FileStreamingOutput(respFile));
-
-			setHeaders(responseBuilder);
-			return responseBuilder.build();
+            return limiter.callWithTimeout(()->{
```

Comment:
IMHO the timeout block could be started from line#486 below. That is where the execution of the queries to C* happen.

Reply:
this timeout is not meant for the cassandra queries...but for the entire response creation process